rfcs/convergence/first-rule-of-aria.md

```diff
@@ -0,0 +1,204 @@
+- Whenever we want to provide options beyond what HTML allows in an interactive widget, we should provide authors an option to use a semantic HTML version (e.g. Dropdown should have an author option to use a `<select>` element, even if we also allow a custom version with a stylable extendable options list).
+- When creating form components, use a `<label for="id">` element instead of `aria-label` or `aria-labelledby`.
+- When creating a component that does not exist in HTML like a tab control, try to make the base interactive controls using semantic HTML elements augmented with ARIA (so `<button role="tab">` instead of `<div role="tab">`) whenever it makes sense to do so.
+
+For example a table following semantic elements would look like:
+
+```tsx
+<Table>
+  <TableHead>
+    <TableRow>
+      <TableCell colSpan={2}>
+        The table header
+      </TableCell>
+    </TableRow>
+  </TableHead>
+  <TableBody>
+    <TableRow>
+      <TableCell>
+        The table body
+      </TableCell>
+      <TableCell>
+        with two columns
+      </TableCell>
+    </TableRow>
+  </TableBody>
+</Table>
+
+
+// renders to
+
+<table>
+  <thead>
+    <tr>
+      <th colspan="2">The table header</th>
+    </tr>
+  </thead>
+  <tbody>
+    <tr>
+      <td>The table body</td>
+      <td>with two columns</td>
+    </tr>
+  </tbody>
+</table>
+```
+
+Since table is a common example of virtualization scenarios, where elements, content, headers, etc,. can be added on the go by JS code, the structural behavior provided by semantic elements can be easily compromised. For example in the case where a div is added between `tbody` and `tr` for scrolling purposes.
+
+```tsx
+<Table>
+  <TableHead>
+    <TableRow>
+      <TableCell colSpan={2}>
+        The table header
+      </TableCell>
+    </TableRow>
+  </TableHead>
+  <TableBody>
+    <div>
+      <TableRow>
+        <TableCell>
+          The table body
+        </TableCell>
+        <TableCell>
+          with two columns
+        </TableCell>
+      </TableRow>
+    </div>
+  </TableBody>
+</Table>
+
+<table>
+  <thead>
+    <tr>
+      <th colspan="2">The table header</th>
+    </tr>
+  </thead>
+  <tbody>
+    <!-- This div breaks the markup, since tbody cannot contain  -->
+    <div>
+      <tr>
+        <td>The table body</td>
+        <td>with two columns</td>
+      </tr>
+    </div>
+  </tbody>
+</table>
+```
+
+The idea of going for semantic first approach is to ensure that in the default cases, where the user of the lib simply doesn't care about edge cases such as `a` element inside of a `button` element, semantic elements will be used ensuring a better overall experience across different environments, and in the edge cases there's an option to dispose the semantic element for something more generic and ARIA.
+
+The example above could easily be converted to ARIA:
+
+```tsx
+// This as="div" is informs all compound components to opt-out of semantic-first approach
+<Table as="div">
```

Comment:
this would have significant impact on styling as div behaves very differently from table.

Reply:
About styling, the default styling from table should be overridden, since semantic elements default styling can't be trusted.
I don't think there should be any difference in styling between a div and a table once the user agente style has been overridden.

Reply:
> this would have significant impact on styling as div behaves very differently from table.

Required markup will be also different, for example:

```html
<div role="table">
  <div role="row">
    <div role="columnheader">id</div>
    <div role="columnheader">Name</div>
  </div>
  <div role="row">
    <div role="cell">1</div>
    <div role="cell">Roman van von der Longername</div>
  </div>
</div>
```

vs.

```html
<table>
  <!-- 👇 is required, without "tbody" we will get a React warning: https://codesandbox.io/s/reverent-shape-iko26 -->
  <!-- 🔴 Warning: validateDOMNesting(...): <tr> cannot appear as a child of <table>. Add a <tbody>, <thead> or <tfoot> ... -->
  <tbody>
    <tr>
      <th>id</th>
      <th>Name</th>
    </tr>
    <tr>
      <td>1</td>
      <td>Roman van von der Longername</td>
    </tr>
  </tbody>
</table>
```

Reply:
`<tbody>` is definitely not a required element in tables (and in some cases would even not be recommended). @layershifter  that sounds like a bug in whatever library or plugin is throwing that warning.

That said, I think any table that requires virtualization, that we want to provide heavy reflowing for at small screens, or grids should probably be marked up with divs + roles.

Reply:
☝️ another good reason why it's preferable to use abstraction closer to the platform we are implementing for - [Preact 🙌](https://preactjs.com/)